gateway/ln-gateway/src/lib.rs

```diff
@@ -1090,52 +1091,98 @@ impl Gateway {
                 password_salt,
             }
         };
-
         dbtx.insert_entry(&GatewayConfigurationKey, &new_gateway_config)
             .await;
+        if let Some(per_federation_routing_fees) = per_federation_routing_fees {
+            self.update_and_register_federation_routing_fees(
+                &mut dbtx,
+                &per_federation_routing_fees,
+                &new_gateway_config,
+            )
+            .await?;
+        }
         dbtx.commit_tx().await;
 
         let mut curr_gateway_config = self.gateway_config.write().await;
         *curr_gateway_config = Some(new_gateway_config.clone());
 
-        self.update_federation_routing_fees(routing_fees, &new_gateway_config)
-            .await?;
         info!("Set GatewayConfiguration successfully.");
 
         Ok(())
     }
 
-    /// Updates the routing fees for every federation configuration. Also
-    /// triggers a re-register of the gateway with all federations.
-    ///
-    /// TODO: Once per-federation fees are supported, this function should only
-    /// update a single federation's routing fees at a time.
-    async fn update_federation_routing_fees(
+    /// Updates the routing fees and register the given federation.
+    async fn update_and_register_federation_routing_fees(
         &self,
-        routing_fees: Option<String>,
+        dbtx: &mut DatabaseTransaction<'_, Committable>,
+        per_federation_routing_fees: &[(FederationId, FederationRoutingFees)],
         gateway_config: &GatewayConfiguration,
     ) -> Result<()> {
-        if let Some(fees_str) = routing_fees {
-            let routing_fees = GatewayFee::from_str(fees_str.as_str())?.0;
-            let mut dbtx = self.gateway_db.begin_transaction().await;
-            let configs = dbtx
-                .find_by_prefix(&FederationIdKeyPrefix)
-                .await
-                .collect::<Vec<_>>()
-                .await;
-
-            for (id_key, mut fed_config) in configs {
-                fed_config.fees = routing_fees;
-                dbtx.insert_entry(&id_key, &fed_config).await;
+        for (federation_id, routing_fees) in per_federation_routing_fees.iter() {
+            let federation_key = FederationIdKey { id: *federation_id };
+            if let Some(mut federation_config) = dbtx.get_value(&federation_key).await {
+                federation_config.fees = routing_fees.clone().into();
+                dbtx.insert_entry(&federation_key, &federation_config).await;
+                if let Err(e) = self
+                    .register_federation(gateway_config, federation_id, &federation_config)
```

Comment:
Now we are re-registering only the federations with new fees. But what about other parameters like `num_route_hints`, will them affect other federations? Should we register again all federations? What do you think @m1sterc001guy ?

Reply:
Yeah, agreed with @douglaz. For other parameters like `num_route_hints`, we should also re-register the gateway with all the federations. 

I think updating the `FederationConfig` and the re-register function can be split apart. Re-register can take a vector of `FederationId`, so we can call it when either `num_route_hints` is updated or when the federation fees are updated.

Eventually with lightning NG we'll be able to get rid of a lot of this complexity.

Reply:
Thanks for catching that! I did the fix as suggested in my last commit.